fs: improve mode and flags validation

BridgeAR opened the pull request:
This fixes a bug in `fs.promises.access` as it accepted strings as
mode. It should have only accepted numbers. It will now always
validate the flags and the mode argument in an consistent way.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

BridgeAR commented:
This mainly fixes the bug but I mark this defensively as semver-major.

CITGM https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1789/

BridgeAR commented:
@nodejs/fs @nodejs/tsc PTAL

BridgeAR commented:
@nodejs/fs @nodejs/tsc PTAL

BridgeAR commented:
This already has some LGs but it needs one more LG from the @nodejs/tsc. PTAL

BridgeAR commented:
I finally got the time to get back to this. The last time I missed a few more bugs in the beginning. That was very subtle because the name of the arguments was deceiving me. I fixed the variable naming and added proper validation for `copyFile`. I also added a couple more tests.

PTAL

BridgeAR commented:
I just pushed two more commits that add some more refinements to the docs and fix all left issues with `mode` and `flag`.

BridgeAR commented:
@Trott @bnoordhuis it would be nice if you could confirm your LG due to the recent changes.

addaleax commented on lib/internal/validators.js:
```diff
@@ -39,7 +39,11 @@ const modeDesc = 'must be a 32-bit unsigned integer or an octal string';
  * @returns {number}
  */
-function parseMode(value, name, def) {
+function parsePermissionMode(value, name, def) {
```
File modes do not just encode permissions. If you want to rename this, I’d suggest `parseFileMode`.

bnoordhuis commented on doc/api/fs.md:
```diff
@@ -5001,16 +4987,16 @@ changes:
-  100ms longer on each try. This option represents the number of retries. This
-  option is ignored if the `recursive` option is not `true`. **Default:** `3`.
+    encountered, Node.js will retry the operation with a linear backoff wait of
+    100ms longer on each try. This option represents the number of retries. This
```
```suggestion
    100 ms longer on each try. This option represents the number of retries. This
```

bnoordhuis commented on doc/api/fs.md:
```diff
@@ -3243,16 +3232,16 @@ changes:
-  100ms longer on each try. This option represents the number of retries. This
-  option is ignored if the `recursive` option is not `true`. **Default:** `3`.
+    encountered, Node.js will retry the operation with a linear backoff wait of
+    100ms longer on each try. This option represents the number of retries. This
```
```suggestion
    100 ms longer on each try. This option represents the number of retries. This
```

bnoordhuis commented on lib/fs.js:
```diff
@@ -175,20 +176,15 @@ function access(path, mode, callback) {
-    mode = F_OK;
-  else
-    mode = mode | 0;
+  mode = getValidMode(mode, 7);
```
This makes too many assumptions about the value of `R_OK|W_OK|X_OK` and also seems to assume that `F_OK == 0` (which it probably is but that's not guaranteed.)

Also applies to line 179.

BridgeAR replied:
I fixed that by returning `F_OK` in case mode is set to `undefined` and by checking for the actual value of `R_OK|W_OK|X_OK|F_OK`.

bnoordhuis commented on lib/fs.js:
```diff
@@ -1786,23 +1786,23 @@ function copyFile(src, dest, flags, callback) {
   src = pathModule._makeLong(src);
   dest = pathModule._makeLong(dest);
-  flags = flags | 0;
+  mode = getValidMode(mode, 7);
```
Magic constant. What does 7 signify?

Also applies to several other uses of `getValidMode()`.

BridgeAR replied:
Fixed by removing the number and instead documenting the behavior in the validation function with a comment.

BridgeAR commented:
@bnoordhuis PTAL

BridgeAR commented:
@cjihrig thanks a lot for the thorough review!

> The changes to the copyFile() functions seem wrong to me. The flags passed to those functions aren't file modes, and shouldn't be validated as such.

Yes, I forgot to distinguish them while addressing another review comment. I changed the implementation to address this. PTAL.

> I think the parameter name flags is more accurate for those (the name was borrowed from the macOS implementation).

I did see that the macOS implementation named them flags and I see the general reasoning there for it. The point why I renamed them in Node.js is to keep it consistent with the rest of our naming. So far it's difficult to see the difference between `flags` and `mode`. I personally got confused when checking the [constants](https://github.com/BridgeAR/node/blob/fix-fs-validation/doc/api/fs.md#fs-constants) and the [flags](https://github.com/BridgeAR/node/blob/fix-fs-validation/doc/api/fs.md#file-system-flags) definition.

It would be great if you could have another look at that part to make a suggestion what you believe is best to prevent the ambiguity. The naming is not the main part for me. It just seems confusing.

Ping @Trott (if you find the time to also check the docs to see how to improve the wording / naming to outline clearer what `flags` and what `mode` stands for in Node.js).

BridgeAR commented:
@cjihrig PTAL. If you think strongly of it, I am fine to take back the name change but it would be great if you could have a closer look first.

I think this should land some time soonish, since it does fix a couple of edge cases.

BridgeAR commented:
Ping @cjihrig

cjihrig commented:
I've dismissed my review, but I'm not going to sign off as I don't really like some of these changes.

BridgeAR commented:
@jasnell @targos @Trott @bnoordhuis please confirm your LG so that this can land.

BridgeAR commented:
One more CITGM https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/2144/  (queued)

BridgeAR commented:
@jasnell @targos @Trott @bnoordhuis please confirm your LG so that this can land.
This PR is open for quite some while.

BridgeAR commented:
Landed in af5ddf21234a4508d7889c8dd172245cf7d4c422...4455f603207815199854797b043b86cc8a885eb3 :tada:
